keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,724 @@
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+1. `bucketClassName`: Name of the associated bucket class.
+1. `bucketRequest`: an `objectReference` containing the name, namespace and UID of the associated `BucketRequest`.
+1. `allowedNamespaces`: a list of namespaces that are permitted to either create new buckets or to access existing buckets.
+1. `protocol`: protocol-specific field the application will use to access the backend storage.
+     - `name`: supported protocols are: "s3", "gs", and "azureBlob".
+     - `version`: version being used by the backend storage.
+     - `azureBlob`: data required to target a provisioned azure container.
+     - `gs`: data required to target a provisioned GS bucket.
+     - `s3`: data required to target a provisioned S3 bucket.
+1. `parameters`: a copy of the BucketClass parameters.
+1. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### BucketClass
+
+An immutable, cluster-scoped, custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a retention policy, driver specific parameters, and the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol. This allows the bucket class to be omitted from a `BucketRequest`. All of the `BucketClass` fields are copied to the `Bucket` instance.  If an object store supports more than one protocol then the admin should create a `BucketClass` per protocol.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+protocol: {"azureblob", "gs", "s3", ... } [3]
+anonymousAccessMode: [4]
+retentionPolicy: {"Delete", "Retain"} [5]
+allowedNamespaces: [6]
+  - name:
+parameters: [7]
+```
+
+1. `provisioner`: (required) the name of the vendor-specific driver supporting the `protocol`.
+1. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then a `BucketRequest` may omit the `BucketClass` reference. If the greenfield `BucketRequest` skips the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used; otherwise an error is logged. It is possible that more than one `BucketClass` of the same protocol is set as the default class. In this case, it is non-deterministic which `BucketClass` is used as the default.
+1. `protocol`: (required) protocol supported by the associated object store. This field validates that the `BucketRequest`'s desired protocol is supported.
```

Comment:
>> Should this also contain version information like BucketRequest?
>
> @jeffvance jeffvance 7 days ago • 
 > Author
> The BC should contain the protocol version as an optional field.
> If we remove BR.protocol then version should also be removed from the BR.

Remember the rule that BC can be deleted and Bucket object must be self-sufficient. Therefore version info seems crucial here. Depending on BR is also not sufficient (if we end up, as I am proposing above, to have a "list of protocols/versions" on BR).

Reply:
BC's protocol should be the same struct as used in the BR. Will update.